### ISSUE TITLE
AQL Query hangs indefinitely

### ISSUE THREAD

**Issue by jtfidje** [arangodb/arangodb#14214]
## My Environment

* __ArangoDB Version__:        3.7.11
* __Storage Engine__:             RocksDB
* __Deployment Mode__:        Cluster
* __Deployment Strategy__:   Manual Start in Docker
* __Used Package__:              Docker - official Docker library

## Steps to reproduce

__Problem__:
If a collection is queried for documents with a `LIMIT` filter and the `fullCount` option set to `true`, the query hangs indefinitely whenever `LIMIT` is lower than the number of documents in the collection.

I've written a description in markdown that two of my teammates used to successfully reproduce the issue on their machines. It includes a docker-compose file to set up the cluster, so for convenience I'll link a PasteBin to it: https://pastebin.com/MQYMbSGD

Our issue looks to be very similar to this as well: https://github.com/arangodb/arangodb/issues/14184

**Comment by chriskvik** [comment]
This is concerning, I'm observing the same behavior in my production cluster on 3.7.11. The queries are hanging and cannot be killed with arangosh. The only way to get rid of them is by restarting the coordinator, but it takes around ~30 minutes because of what I'm assuming is a timeout on "killing running searches". 

I'm constantly monitoring my production setup, so that this wont knock out all my coordinator, while restarting them continuously.  

My current workaround is to ensure no query asks for a collection count LIMIT that is lower than the collection count. But this one is nasty.

**Comment by chriskvik** [comment]
After further investigation we found that using `SORT` also alleviates the issue.
```
FOR document in some_collection
    SORT document.some_property
    LIMIT 0, 1
    RETURN document
```

**Comment by dothebart** [comment]
can you also add an explain output of this query?

**Comment by chriskvik** [comment]
Sure @dothebart. Below is the output of `/_api/explain` for the query with `SORT`  (working).

```
{
  "plan": {
    "nodes": [
      {
        "type": "SingletonNode",
        "dependencies": [],
        "id": 1,
        "estimatedCost": 1,
        "estimatedNrItems": 1
      },
      {
        "type": "EnumerateCollectionNode",
        "dependencies": [
          1
        ],
        "id": 2,
        "estimatedCost": 14,
        "estimatedNrItems": 12,
        "random": false,
        "indexHint": {
          "forced": false,
          "type": "none"
        },
        "outVariable": {
          "id": 0,
          "name": "document",
          "isDataFromCollection": true
        },
        "projections": [],
        "count": false,
        "producesResult": true,
        "database": "some_database",
        "collection": "some_collection",
        "satellite": false,
        "numberOfShards": 9
      },
      {
        "type": "CalculationNode",
        "dependencies": [
          2
        ],
        "id": 3,
        "estimatedCost": 26,
        "estimatedNrItems": 12,
        "expression": {
          "type": "attribute access",
          "typeID": 35,
          "name": "name",
          "subNodes": [
            {
              "type": "reference",
              "typeID": 45,
              "name": "document",
              "id": 0
            }
          ]
        },
        "outVariable": {
          "id": 2,
          "name": "1",
          "isDataFromCollection": false
        },
        "canThrow": false,
        "expressionType": "attribute"
      },
      {
        "type": "SortNode",
        "dependencies": [
          3
        ],
        "id": 4,
        "estimatedCost": 69.01955000865388,
        "estimatedNrItems": 12,
        "elements": [
          {
            "inVariable": {
              "id": 2,
              "name": "1",
              "isDataFromCollection": false
            },
            "ascending": true
          }
        ],
        "stable": false,
        "limit": 0,
        "strategy": "standard"
      },
      {
        "type": "RemoteNode",
        "dependencies": [
          4
        ],
        "id": 9,
        "estimatedCost": 81.01955000865388,
        "estimatedNrItems": 12,
        "distributeId": "",
        "isResponsibleForInitializeCursor": true,
        "database": "some_database",
        "server": "",
        "queryId": "",
        "api": 1
      },
      {
        "type": "GatherNode",
        "dependencies": [
          9
        ],
        "id": 10,
        "estimatedCost": 93.01955000865388,
        "estimatedNrItems": 12,
        "parallelism": "parallel",
        "sortmode": "heap",
        "limit": 1,
        "elements": [
          {
            "inVariable": {
              "id": 2,
              "name": "1",
              "isDataFromCollection": false
            },
            "ascending": true
          }
        ]
      },
      {
        "type": "LimitNode",
        "dependencies": [
          10
        ],
        "id": 5,
        "estimatedCost": 94.01955000865388,
        "estimatedNrItems": 1,
        "offset": 0,
        "limit": 1,
        "fullCount": true
      },
      {
        "type": "ReturnNode",
        "dependencies": [
          5
        ],
        "id": 6,
        "estimatedCost": 95.01955000865388,
        "estimatedNrItems": 1,
        "inVariable": {
          "id": 0,
          "name": "document",
          "isDataFromCollection": true
        },
        "count": true
      }
    ],
    "rules": [
      "scatter-in-cluster",
      "distribute-filtercalc-to-cluster",
      "distribute-sort-to-cluster",
      "remove-unnecessary-remote-scatter",
      "sort-limit",
      "parallelize-gather"
    ],
    "collections": [
      {
        "name": "some_collection",
        "type": "read"
      }
    ],
    "variables": [
      {
        "id": 2,
        "name": "1",
        "isDataFromCollection": false
      },
      {
        "id": 0,
        "name": "document",
        "isDataFromCollection": true
      }
    ],
    "estimatedCost": 95.01955000865388,
    "estimatedNrItems": 1,
    "isModificationQuery": false
  },
  "cacheable": true,
  "warnings": [],
  "stats": {
    "rulesExecuted": 52,
    "rulesSkipped": 0,
    "plansCreated": 1
  },
  "error": false,
  "code": 200
}
```

And the output for `/_api/explain` for the failing query without `SORT`.

```
{
  "plan": {
    "nodes": [
      {
        "type": "SingletonNode",
        "dependencies": [],
        "id": 1,
        "estimatedCost": 1,
        "estimatedNrItems": 1
      },
      {
        "type": "EnumerateCollectionNode",
        "dependencies": [
          1
        ],
        "id": 2,
        "estimatedCost": 14,
        "estimatedNrItems": 12,
        "random": false,
        "indexHint": {
          "forced": false,
          "type": "none"
        },
        "outVariable": {
          "id": 0,
          "name": "document",
          "isDataFromCollection": true
        },
        "projections": [],
        "count": false,
        "producesResult": true,
        "database": "some_database",
        "collection": "some_collection",
        "satellite": false,
        "numberOfShards": 9
      },
      {
        "type": "RemoteNode",
        "dependencies": [
          2
        ],
        "id": 7,
        "estimatedCost": 26,
        "estimatedNrItems": 12,
        "distributeId": "",
        "isResponsibleForInitializeCursor": true,
        "database": "some_database",
        "server": "",
        "queryId": "",
        "api": 1
      },
      {
        "type": "GatherNode",
        "dependencies": [
          7
        ],
        "id": 8,
        "estimatedCost": 38,
        "estimatedNrItems": 12,
        "parallelism": "parallel",
        "sortmode": "unset",
        "elements": []
      },
      {
        "type": "LimitNode",
        "dependencies": [
          8
        ],
        "id": 3,
        "estimatedCost": 39,
        "estimatedNrItems": 1,
        "offset": 0,
        "limit": 1,
        "fullCount": true
      },
      {
        "type": "ReturnNode",
        "dependencies": [
          3
        ],
        "id": 4,
        "estimatedCost": 40,
        "estimatedNrItems": 1,
        "inVariable": {
          "id": 0,
          "name": "document",
          "isDataFromCollection": true
        },
        "count": true
      }
    ],
    "rules": [
      "scatter-in-cluster",
      "remove-unnecessary-remote-scatter",
      "parallelize-gather"
    ],
    "collections": [
      {
        "name": "some_collection",
        "type": "read"
      }
    ],
    "variables": [
      {
        "id": 0,
        "name": "document",
        "isDataFromCollection": true
      }
    ],
    "estimatedCost": 40,
    "estimatedNrItems": 1,
    "isModificationQuery": false
  },
  "cacheable": true,
  "warnings": [],
  "stats": {
    "rulesExecuted": 52,
    "rulesSkipped": 0,
    "plansCreated": 1
  },
  "error": false,
  "code": 200
}
```

**Comment by dothebart** [comment]
I was rather thinking about the formated one, which is more human readable ;-) but thanks.

**Comment by chriskvik** [comment]
> I was rather thinking about the formated one, which is more human readable ;-) but thanks.

Can I export the human readable one via the API? As far as I know, it's not possible to set `fullCount: true` (required to reproduce the issue) using the AQL Query editor.

**Comment by dothebart** [comment]
in arangosh you can run `db._explain(query)` - or you can click explain in the webinterface in the query tab.

**Comment by chriskvik** [comment]
Here you go!

**Not working.**

```
some_database:8529@some_database> db._explain({query:'FOR some_object in some_collection LIMIT 10 RETURN some_object', options: {'fullCount': true}})
Query String (56 chars, cacheable: true):
 FOR some_object in some_collection LIMIT 10 RETURN some_object

Execution plan:
 Id  NodeType                 Site Est.  Comment
 1  SingletonNode            DBS     1  * ROOT
 2  EnumerateCollectionNode  DBS    13    - FOR some_object IN some_collection  /* full collection scan, 9 shard(s) /
 7  RemoteNode               COOR   13      - REMOTE
 8  GatherNode               COOR   13      - GATHER  / parallel, unsorted /
 3  LimitNode                COOR   10      - LIMIT 0, 10 / fullCount */
 4  ReturnNode               COOR   10      - RETURN some_object

Indexes used:
 none

Optimization rules applied:
 Id  RuleName
 1  scatter-in-cluster
 2  remove-unnecessary-remote-scatter
 3  parallelize-gather
```
**Working.**

```
some_database:8529@some_database> db._explain({query:'FOR document in collection_w_one_doc FOR some_object in some_collection LIMIT 10 RETURN some_object', options: {'fullCount': true}})
Query String (74 chars, cacheable: true):
 FOR document in collection_w_one_doc FOR some_object in some_collection LIMIT 10 RETURN some_object

Execution plan:
 Id  NodeType                 Site Est.  Comment
 1  SingletonNode            DBS     1  * ROOT
 14  IndexNode                DBS     1    - FOR document IN collection_w_one_doc  /* primary index scan, scan only, 9 shard(s) /   
 12  RemoteNode               COOR    1      - REMOTE
 13  GatherNode               COOR    1      - GATHER  / unsorted /
 6  ScatterNode              COOR    1      - SCATTER
 7  RemoteNode               DBS     1      - REMOTE
 3  EnumerateCollectionNode  DBS    13      - FOR some_object IN some_collection  / full collection scan, 9 shard(s) /
 8  RemoteNode               COOR   13        - REMOTE
 9  GatherNode               COOR   13        - GATHER  / unsorted /
 4  LimitNode                COOR   10        - LIMIT 0, 10 / fullCount */
 5  ReturnNode               COOR   10        - RETURN some_object

Indexes used:
 By  Name     Type     Collection  Unique  Sparse  Selectivity  Fields      Ranges
 14  primary  primary  collection_w_one_doc       true    false      100.00 %  [ _key ]  *

Optimization rules applied:
 Id  RuleName
 1  scatter-in-cluster
 2  remove-unnecessary-remote-scatter
 3  reduce-extraction-to-projection
```

**Comment by mpoeter** [comment]
I am able to reproduce this locally - investigating!

**Comment by mpoeter** [comment]
This should be fixed via the mentioned PRs (see above). Until a release with the fix is available you can work around this problem by disabling the `parallelize-gather` rule for the query:
```
db._query({
  query: 'FOR some_object in some_collection LIMIT 10 RETURN some_object',
  options: {
    fullCount: true,
    optimizer: { rules: ["-parallelize-gather"] }
  }
})
```

**Comment by maxkernbach** [comment]
Hi @jtfidje and @chriskvik,

In the meantime version 3.7.14 has been released. A fix for the case of indefinitely hanging queries is already available in 3.7.13.
Could you please give it a try?

**Comment by Simran-B** [comment]
@jtfidje @chriskvik Is the problem resolved with the latest versions?
